### PR TITLE
Integrate Scroll prover with Snarkify SDK

### PR DESCRIPTION
**Pull request by Gigatron**

### CODE REVIEW

File: prover/src/bin/snarkify.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        let prover = Prover::new(&config).map_err(|e| e.to_string())?;
+
+        log::info!(
+            "prover start successfully. name: {}, type: {:?}, publickey: {}, version: {}",
+            config.prover_name,
+            config.proof_type,
+            prover.get_public_key(),
+            version::get_version(),
+        );
```
Comment (Gigatron) [inline prover/src/bin/snarkify.rs]: @cyphersnake, Need some advice here. We want the prover to be reused across different reqeusts, so we woud like it to be stored as a static variable and lazily initialized. Chatgpt recommended to use 
```
static PROVER: OnceCell<Prover> = OnceCell::new();

let prover = PROVER.get_or_init(|| {
  ...
})

```

but the compiler is giving many errors about thread-safty. In our use case we know that it is a single-threaded model, how should I rewrite the code to achieve my goal. Can you help take a look?

Comment (cyphersnake) [reply prover/src/bin/snarkify.rs]: https://doc.rust-lang.org/std/macro.thread_local.html

```rust
use std::cell::RefCell;

struct Prover;

impl Prover {
    fn new() -> Self {
        Self
    }

    fn ping(&self) -> u8 {
        42
    }

    fn pong(&mut self) -> u8 {
        42
    }
}

thread_local! {
    static GLOBAL: RefCell<Prover> = RefCell::new(Prover::new());
}

fn main() {
    GLOBAL.with_borrow(|v| {
        v.ping();
    });

    GLOBAL.with_borrow_mut(|v| {
        v.pong();
    });
}
```
Let me know if that's not enough

Comment (Gigatron) [reply prover/src/bin/snarkify.rs]: Made the change as you suggested, and I think it works! Thanks!

File: prover/Cargo.toml [hunk 2]
```diff
@@ -10,12 +10,13 @@ edition = "2021"
...
 [patch."https://github.com/privacy-scaling-explorations/halo2.git"]
-halo2_proofs = { git = "https://github.com/scroll-tech/halo2.git", branch = "v1.1"  }
+halo2_proofs = {  git = "https://github.com/snarkify/halo2-scroll.git", tag = "snarkify-v0.1.5" }
 [patch."https://github.com/privacy-scaling-explorations/poseidon.git"]
 poseidon = { git = "https://github.com/scroll-tech/poseidon.git", branch = "main" }
 [patch."https://github.com/privacy-scaling-explorations/bls12_381"]
 bls12_381 = { git = "https://github.com/scroll-tech/bls12_381", branch = "feat/impl_scalar_field" }
-
+[patch."https://github.com/scroll-tech/halo2.git"]
+halo2_proofs = {  git = "https://github.com/snarkify/halo2-scroll.git", tag = "snarkify-v0.1.5" }
```
Comment (ErdongChen61) [inline prover/Cargo.toml:19]: nit: remove this line as it is the same as line 13

File: prover/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# The build step is skpped, one should build on the host first
```
Comment (ErdongChen61) [inline prover/Dockerfile]: nit: `skipped`

File: prover/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+## How-to run Snarkify prover locally
+1. `make snarkify`
```
Comment (ErdongChen61) [inline prover/README.md:2]: what does it mean? install the snarkify cli?

Comment (Gigatron) [reply prover/README.md:2]: no this is a build target in the Makefile to build the snarkify binary from Rust code

File: prover/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+2. update `config.json` with right coordinator and l2geth endpoints
```
Comment (ErdongChen61) [inline prover/README.md]: nit: better to list what are the endpoints for testnet and mainnet.

Comment (ErdongChen61) [reply prover/README.md]: and why prover needs coordinator endpoint? My gut is that only relayer needs it.

**Review by ErdongChen61 — APPROVED** [review]

File: prover/README.md (hunk 5 above)
Comment (Gigatron) [reply prover/README.md]: It doesn't use coordinator endpoint, I'll remove this requirement for now.

**Pull request merged**